### ISSUE TITLE
Installability: Page does not work offline

### ISSUE THREAD

**Issue by FadedOut** [monsoonconsulting/magento2-pwa#5]
Hello,

First want to say much appreciation of you bringing this to us Magento users. It was a cinch to install and the PWA does indeed work. Not sure why the major complicated hassle required with the "PWA Studio" (creating a separate theme, installing things like node, etc. etc.) when all that is required really is a service worker. 

Unless, my issue stems from this _not_ being the "PWA Studio" in which case will be disappointing to say the least (as I can't figure out PWA studio and don't have time to).

My issue: as the title states, in Chrome DevTools under the Application tab it says "Installability: Page does not work offline". I resolved a few other errors like id missing (just added to the manifest.json file) and changing the "any maskable" to just "maskable" per web.dev suggestions. But the page not working offline error is stumping me. I have all settings to cache in the Magento admin (Cache Static Assets: Yes, Cache Media Assets: Yes, Cache Google Fonts: Yes. I left "Enable Skip Waiting" set to no, and "Enable clients.claim" set to no). 

Is there any particular step needed to make it work offline? I don't get the little "Add to home" banner popup in the mobile browser (though clicking the three dots then clicking "Install" does work) nor do I get the little "Install" icon in the desktop chrome URL bar that normally appears if the PWA works properly. Also if I disable network/wifi on my phone, the PWA does not work - it says offline. For the "Offline page" I chose my homepage. 

Any help with this would be super appreciated. And thanks again for bringing this as an alternative to the PWA Studio.

**Comment by rossmc** [comment]
Hi @FadedOut,

Thanks for the feedback, you are correct, you don't need PWA studio to make Magento a PWA. That was the main reason behind creating this module.

Regarding your issue _"Installability: Page does not work offline"_.
Are you serving your site over HTTP or HTTPS? 
PWAs only work over HTTPS, and I believe this is the error you see if you are serving over HTTP.
REF:
https://love2dev.com/blog/is-your-progressive-web-app-may-not-prompting-the-customer-to-add-to-home-screen/
https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps/Offline_Service_workers#security

For it to work offline you need to enable the `Offline Notification Page`
![CleanShot 2022-10-22 at 22 00 55@2x](https://user-images.githubusercontent.com/2452991/197360586-6a18f6d1-8ddc-47fd-ac24-e4a1bab842ab.png)

I tested this in a fresh Magento install and could not repicate, it none of the above is helpful, please provide more detail, browser console errors, screenshots of application tab, and please run a chrome lighthouse PWA report and share a screenshot of the results. And I'll try to help as best I can.

<img width="300" src="https://user-images.githubusercontent.com/2452991/197360695-4e418eab-4f80-4dc5-bced-59a9067d0749.png">

<img width="300" src="https://user-images.githubusercontent.com/2452991/197360717-1dfed9f2-641e-443c-bd27-01c17100075b.png">

<img width="300" src="https://user-images.githubusercontent.com/2452991/197360732-59e8a145-05f2-42ee-8759-3423c967fc5c.png">

<img width="300" src="https://user-images.githubusercontent.com/2452991/197360733-7d002622-e459-4d06-adaf-dc93f116c5de.png">

**Comment by rossmc** [comment]
Ahh, I see you said "For the "Offline page" I chose my homepage."

What happens if you set the Offline page which comes with this module?

**Comment by FadedOut** [comment]
Hi @rossmc 

Thanks for getting back to me it's much appreciated!

I am indeed serving over HTTPS and I'm using Magento 2.4.5p1

Here is the lighthouse screenshot: https://imgur.com/a/RjLsg64

Also the backend settings: https://imgur.com/a/anXDiv9

Here is a screenshot of my manifest file: https://imgur.com/a/tM8YoAZ

Screenshot of the Application tab: https://imgur.com/a/AH67jIo

I had a blockage of the service-worker from the CSP but I added the exception for it and now the console does not show any error blocking it. So no console logs pertaining to the pwa. 

I just changed to the "Offline" page that is included with the module and it made no change - chrome still says it's not installable. 

And I made sure to clear browser cache. But also on my actual phone it's not "popping up" the add to homescreen - though as mentioned if I click the three dots (menu) there is a "Install" option which if clicked does add to the home screen and does work as a PWA - albeit not offline (going into airplane mode the PWA shows just an offline no network error). 

Hopefully it's just some oversight on my part and is a simple resolution so it doesn't take up much of your time. 

Thank you in advance!

**Comment by rossmc** [comment]
Thanks @FadedOut,

It seems like the service worker is not registering.

What do you see in the application tab's service worker menu?
<img width="200" alt="CleanShot 2022-10-23 at 00 28 39@2x" src="https://user-images.githubusercontent.com/2452991/197364596-8b2e03c0-e931-473f-bd95-8e58a589d5d9.png">

Can you check if the service worker file exists on your site, go to `[your domain]/service-worker.js` in the browser to check.
<img width="200" alt="CleanShot 2022-10-23 at 00 24 32@2x" src="https://user-images.githubusercontent.com/2452991/197364653-c60db581-d6d3-4e58-bf95-b89dac49df4a.png">

Finally, which theme are you using, what is it's parent theme? If your theme is not Luma/Hyva based this could be the cause. 

If your theme is luma based you should be able to see a registration.js file being loaded in the browser?
<img width="200" alt="CleanShot 2022-10-23 at 00 35 57@2x" src="https://user-images.githubusercontent.com/2452991/197364733-dd9bbc6d-e545-40db-aff6-d50afa767eb7.png">

And if your theme is hyva based, you should be able to find the text `if ('serviceWorker' in navigator)` if you view the source of the webpage.

Thanks,
Ross

**Comment by FadedOut** [comment]
It shows it is available and working I believe: https://imgur.com/a/LAX1uMh

It also shows (the file code) if I go to my-domain.com/service-worker.js

I'm using the "Codazon - Infinit" theme which says it is based on Luma - they say anything that works for luma should work for this theme. 

If it would help, you might be able to diagnose a little easier by using the dev tools on my site here: https://bit.ly/3TJnJ6T (masked for bots) it might help save some time in pinpointing what I'm doing wrong (or if it's something else). 

Thank you 🙂

**Comment by rossmc** [comment]
Thanks @FadedOut but I can't access the site. Seems you have a CF rule blocking me.
<img width="200" src="https://user-images.githubusercontent.com/2452991/197383033-b9a42055-1085-4a9d-80c7-480b1a0c0805.png" />

**Comment by FadedOut** [comment]
Ah shoot, where are you located (country)? I'll temp unblock it

**Comment by rossmc** [comment]
Hi @FadedOut ,

I was able to access with a VPN.

Found the issue, its is due a lot of the code in the [service-worker.js](https://github.com/monsoonconsulting/magento2-pwa/blob/master/view/frontend/templates/service-worker.js.phtml) file being commented out.

<img width="300" src="https://user-images.githubusercontent.com/2452991/197401930-82b7e991-a6cd-4d6b-8733-94aeef2613e6.png" />

This could be due to JS minification on [Magento](https://docs.magento.com/user-guide/configuration/advanced/developer.html#javascript-settings) or [Cloudflare](https://support.cloudflare.com/hc/en-us/articles/200168196-Using-Cloudflare-Auto-Minify).      
What settings do you have for Cloudflare?     
If they're enabled, please disable "Auto Minify" and "Rocket Loader".  
_(FYI: minification on Cloudflare should not be used if it is also being used on Magento, and should never be used for Magento's HTML)_

Tomorrow, I'll try to reproduce on Magento 2.4.5p1, maybe it's Magento's own JavaScript minification causing the issue.

Thanks,
Ross

**Comment by FadedOut** [comment]
Hello @rossmc 

Thank you for continuing to try and investigate this. I really appreciate it. 

I did have minification as well as merge files on through Magento only. I do not use Cloudflare's minification at all (to use Cloudflare's is not usually a good idea with big complicated scripts like this as it's going to be too aggressive and do more harm than good). 

![cloudflare-minify-off](https://user-images.githubusercontent.com/12685113/197416259-1efb67f3-27b3-49b0-a9c3-ccf06c46fcc4.jpg)

So through Magento, I tried turning off minification then tested, didn't seem to make any change to that file. I then tried turning off merge, that also didn't seem to make any difference. I used these commands:

`bin/magento config:set dev/js/merge_files 0 --lock-env`
`bin/magento config:set dev/js/minify_files 0 --lock-env`

I'm also making sure to run this AIO command [that I always use] after every change (and I clear Cloudflare Cache):

`rm -rf var/cache/* ; rm -rf var/page_cache/* ; rm -rf var/view_preprocessed/* ; rm -rf pub/static/*/* ; rm -rf generated/code/* ; rm -rf generated/metadata/* ; php bin/magento setup:upgrade ; php bin/magento setup:static-content:deploy ; php bin/magento cache:flush ; php bin/magento cache:clean`

You are correct though; something is commenting out basically that entire file. Which I think would indicate it has to be something specific with my site then? Herm.

**Comment by rossmc** [comment]
Hi @FadedOut ,

As you say it does seem to be related to your site, I was unable to reproduce it.

One thing I noticed when testing locally, is that with minify_html on it did minify the service-worker (as it is added with a template file). However it was minified without any errors or commenting out code, as should expected with Magento [Minifier.php](https://github.com/magento/magento2/blob/2.4.5-p1/lib/internal/Magento/Framework/View/Template/Html/Minifier.php#L141). Still, just incase, I would try to disable HTML minifcation on Magento.

`bin/magento config:set dev/template/minify_html 0 --lock-env`

If that doesn't work try removing all JavaScript comments from this module's `view/frontend/templates/service-worker.js.phtml` file. You can override it in your theme.

Next, check if there are any messages in the Magento logs on the file system at `var/log/`, maybe there is a clue there.

Finally, I would check which modules you have installed. Are any of them related to performance enchantments, maybe one of the theme modules does? Check their config options or try to disable them temporarily.

Hope some of this helps 🤞 

Kind Regards,
Ross

**Comment by rossmc** [comment]
@FadedOut I release a [small improvement](https://github.com/monsoonconsulting/magento2-pwa/releases/tag/1.1.1) which may help you here. By replacing the single line comments, with multi-line comments. If the minifier leaves the comments in, it should not comment out the code.

You can update with:

`composer require monsoonconsulting/magento2-pwa:1.1.1`

**Comment by FadedOut** [comment]
> Hi @FadedOut ,
> 
> As you say it does seem to be related to your site, I was unable to reproduce it.
> 
> One thing I noticed when testing locally, is that with minify_html on it did minify the service-worker (as it is added with a template file). However it was minified without any errors or commenting out code, as should expected with Magento [Minifier.php](https://github.com/magento/magento2/blob/2.4.5-p1/lib/internal/Magento/Framework/View/Template/Html/Minifier.php#L141). Still, just incase, I would try to disable HTML minifcation on Magento.
> 
> `bin/magento config:set dev/template/minify_html 0 --lock-env`
> 
> If that doesn't work try removing all JavaScript comments from this module's `view/frontend/templates/service-worker.js.phtml` file. You can override it in your theme, as described in [Configure Service Worker](https://github.com/monsoonconsulting/magento2-pwa#configure-service-worker) from the README.
> 
> Next, check if there are any messages in the Magento logs on the file system at `var/log/`, maybe there is a clue there.
> 
> Finally, I would check which modules you have installed. Are any of them related to performance enchantments, maybe one of the theme modules does? Check their config options or try to disable them temporarily.
> 
> Hope some of this helps 🤞
> 
> Kind Regards, Ross

Yay 👍👍👍

You're good 🙂

I disabled the minify_html and that didn't do it. But then I did as you suggested and removed all comment lines from the js file.....THAT did it 👍😀

I re-enabled the minify_html, css minify, js minify as well as merge_files and it still works 😀👍

Dev tools does not show "Page does not work offline" error anymore: https://imgur.com/X9Q6kCZ

And it is showing installable in the browser bar for chrome: https://imgur.com/ImAc6Jd

Super awesome, very happy! Odd that those lines were causing an issue for me but not on your clean test install...that is weird. It also didn't seem to have anything to do with minifying or merging files. Just those comment lines. 

In any case, I super appreciate the time spent on this. Do you have a donate link (like through paypal) that I could send a few dollars to in the near future for the help and great work on this module? I have to wait till I get paid but I'd like to send a donation when I can. 

Thanks again Ross, great work 🙂👍

**Comment by FadedOut** [comment]
> @FadedOut I release a [small improvement](https://github.com/monsoonconsulting/magento2-pwa/releases/tag/1.1.1) which may help you here. By replacing the single line comments, with multi-line comments. If the minifier leaves the comments in, it should not comment out the code.
> 
> You can update with:
> 
> `composer require monsoonconsulting/magento2-pwa:1.1.1`

I will have to try this out and see if it works without modifying the file as I just did (by removing the comment lines). Will report back

**Comment by FadedOut** [comment]
> @FadedOut I release a [small improvement](https://github.com/monsoonconsulting/magento2-pwa/releases/tag/1.1.1) which may help you here. By replacing the single line comments, with multi-line comments. If the minifier leaves the comments in, it should not comment out the code.
> 
> You can update with:
> 
> `composer require monsoonconsulting/magento2-pwa:1.1.1`

Yup this worked 👍

I guess it didn't like the / for comment out and preferred the /* type. Still odd how it didn't give you problems. In any case, much appreciate helping me get it up and running. Even after all that, STILL much easier than the "official" PWA Studio that Magento touts lol. They're crazy for doing so. 

If possible, I'd like to make 1 request, hope you don't mind. Would it be possible to have a "X" or "close" button on the "Add to Home Screen" button that pops up then maybe have a cookie attached? As it is right now, it's kind of in the customers face to install. They may not want to and it would make them happy if they could close the button so it's not harassing them to "Add to home screen" on every page. I'm not sure how much work that would be to implement, I'm hoping not much? 

Still a great module and still appreciate the support!

**Comment by rossmc** [comment]
Great, I'm glad it worked :) 

Yeah strange it was not occurring on a fresh install, maybe some customisation from another module on your site. Anyway I'm very happy it work for you, these style of comments are much better here, easier to read and less likely to cause an issue with a Minifier. Should have used them in the first place. Thank you for raising.

Good suggestion regarding the button, can you create a new issue (feature request) for it please? I'll add it when I have time.
